Modules/_json.c

```diff
@@ -1535,84 +1592,35 @@ encoder_listencode_dict(PyEncoderObject *s, _PyUnicodeWriter *writer,
         */
     }
 
-    items = PyMapping_Items(dct);
-    if (items == NULL)
-        goto bail;
-    if (s->sort_keys && PyList_Sort(items) < 0) {
-        Py_DECREF(items);
-        goto bail;
-    }
-    it = PyObject_GetIter(items);
-    Py_DECREF(items);
-    if (it == NULL)
-        goto bail;
-    idx = 0;
-    while ((item = PyIter_Next(it)) != NULL) {
-        PyObject *encoded, *key, *value;
-        if (!PyTuple_Check(item) || PyTuple_GET_SIZE(item) != 2) {
-            PyErr_SetString(PyExc_ValueError, "items must return 2-tuples");
+    if (s->sort_keys) {
+        Py_ssize_t i;
+
+        items = PyDict_Items(dct);
```

Comment:
> Leave the old path with PyMapping_Items that creates a list of
key-value tuples for cases when sorting is requested.

According to your suggestion: 
```suggestion
        items = PyMapping_Items(dct);
```

But looks not, please let me know if I misunderstood.

Reply:
Yes, that was an outdated commit description. I put `PyMapping_Items` back (instead of `PyDict_Items`) because there's no significant difference. But I kept the `PyList_GET_ITEM` instead of `PyIter_Next` because we call `PyList_Sort` on `items` and we know we have a list. And avoiding iterator leads to fewer references to keep track of: one for the iterator itself and then for each return value of PyIter_Next.